### PR TITLE
KIALI-1704 Verify DR subsets point to a real version

### PR DESCRIPTION
**Pull request by burmanm**
Currently we only test that the target host exists. This adds subset checking to the DR verification, if subsets have been defined in the destination rule.

Also, changes service fetching to be async.

### CODE REVIEW

**Comment by rhqci** [comment]
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#260](https://paste.fedoraproject.org/paste/8941ZHS1o-tB2Xz0DjJe6A)
  * :heavy_check_mark: Istio-and-Kiali-on-Openshift [#[1779]](https://paste.fedoraproject.org/paste/fUNUY9qa060HTr-pyOW72Q)
  * :heavy_check_mark: Istio-Upstream [#[624]](https://paste.fedoraproject.org/paste/6uUVf~ppMwqUDZ1E~m7hRw)
  * :heavy_check_mark: Kiali-Istio-Checker [#[2106, 2107]](https://paste.fedoraproject.org/paste/olERCxZG9TkLbjo8qhhgkQ)
  * :heavy_check_mark: install-bookinfo [#[1154]](https://paste.fedoraproject.org/paste/jmUjMmcJ6gvwEimfy2EF5A)
  * :heavy_check_mark: Kiali-on-Openshift [#[1148]](https://paste.fedoraproject.org/paste/21SwDkEhlFLnO79HIr-Y3w)
  * :heavy_check_mark: run-kiali-e2e-tests [#[551]](https://paste.fedoraproject.org/paste/P7etpsKKU04SbElINnh1pw)


KIALI_NAME=kialicorepr596-istio-system

File: business/checkers/destination_rules/no_host_checker.go [hunk 1]
```diff
@@ -1,33 +1,57 @@
...
 
 type NoHostChecker struct {
 	Namespace       string
-	ServiceNames    []string
+	Services        map[string][]string
 	DestinationRule kubernetes.IstioObject
 }
 
-func (destinationRule NoHostChecker) Check() ([]*models.IstioCheck, bool) {
+// Check parses the DestinationRule definitions and verifies that they point to an existing service, including any subset definitions
```
Comment (lucasponce) [inline business/checkers/destination_rules/no_host_checker.go]: This is good. I wonder about the naming, as we check host/version == destination pair, I think no_destination_checker.go sounds more meaningful, wdyt ? but it's minor.

File: business/istio_validations.go [hunk 2]
```diff
@@ -47,33 +50,35 @@ func (in *IstioValidationsService) GetServiceValidations(namespace, service stri
 
 func (in *IstioValidationsService) GetNamespaceValidations(namespace string) (models.NamespaceValidations, error) {
 	// Get all the Istio objects from a Namespace
+	// TODO This isn't getting all of them, only blocking processing so we can fetch rest later. Needs refactoring
```
Comment (lucasponce) [inline business/istio_validations.go]: I agree.
It's clear we can't continue doing individual operations per element, as we need at least info from the entire namespace to validate an element.

**Comment by lucasponce** [comment]
Some comments added in the PR.
@hhovsepy could you verify the bug is gone ? Thanks.

**Comment by rhqci** [comment]
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#277](https://paste.fedoraproject.org/paste/~EJpcYGlmIcMNsKvMhtL9w)
  * :heavy_check_mark: Istio-and-Kiali-on-Openshift [#[1806]](https://paste.fedoraproject.org/paste/yIwlAp16vIjMgcEv57FhlQ)
  * :heavy_check_mark: Istio-Upstream [#[647]](https://paste.fedoraproject.org/paste/2zUw-~2bULvqlOXJhMBRlw)
  * :heavy_check_mark: Kiali-Istio-Checker [#[2160, 2162]](https://paste.fedoraproject.org/paste/rWxVOt-IOdz-4SDYmo5~Bw)
  * :heavy_check_mark: install-bookinfo [#[1181]](https://paste.fedoraproject.org/paste/A7803RI0f8u3ZfUONONvCQ)
  * :heavy_check_mark: Kiali-on-Openshift [#[1179]](https://paste.fedoraproject.org/paste/rsuRmh5mWFc~YBbfkL9v7A)
  * :heavy_check_mark: run-kiali-e2e-tests [#[580]](https://paste.fedoraproject.org/paste/k~6hxVbvW2bim9Ncqg~Sig)


KIALI_NAME=kialicorepr596-istio-system

**Comment by hhovsepy** [comment]
It shows an error on unavailable version.
But the number of errors in istio config list hover-over is shown 3, and in details page it is 1.

screenshot from 2018-10-29 09-52-50

screenshot from 2018-10-29 09-52-53

File: business/checkers/no_service_checker.go [hunk 3]
```diff
@@ -93,12 +98,30 @@ func runDestinationRuleCheck(destinationRule kubernetes.IstioObject, namespace s
...
+}
+
 func getServiceNames(services []v1.Service) []string {
 	serviceNames := make([]string, 0)
-	if services != nil {
-		for _, item := range services {
-			serviceNames = append(serviceNames, item.Name)
-		}
+	for _, item := range services {
+		log.Infof("Service %s labels: %v, annotations: %v\n", item.Name, item.Labels, item.Annotations)
```
[the remark this answers is not in the document]
Comment (burmanm) [reply business/checkers/no_service_checker.go]: .. these were not supposed to be here ;)

**Comment by rhqci** [comment]
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#278](https://paste.fedoraproject.org/paste/2~FrnQXXqY06-XjwzIWuBQ)
  * :heavy_check_mark: Istio-and-Kiali-on-Openshift [#[1807]](https://paste.fedoraproject.org/paste/c1tUDhPDQavj5HlILAaGbg)
  * :heavy_check_mark: Istio-Upstream [#[648]](https://paste.fedoraproject.org/paste/23zl4xVZv-cXb2MvEDn2YQ)
  * :heavy_check_mark: Kiali-Istio-Checker [#[2163, 2164]](https://paste.fedoraproject.org/paste/y0VckwjArCvO75ra1jCmjQ)
  * :heavy_check_mark: install-bookinfo [#[1183]](https://paste.fedoraproject.org/paste/ZE3DVDSFLeffox88fqTAvw)
  * :heavy_check_mark: Kiali-on-Openshift [#[1180]](https://paste.fedoraproject.org/paste/AMqP7HH0k1h5eJn8lj4YYQ)
  * :heavy_check_mark: run-kiali-e2e-tests [#[581]](https://paste.fedoraproject.org/paste/XAO9K67BMezazKEhBTWwXQ)


KIALI_NAME=kialicorepr596-istio-system

**Comment by burmanm** [comment]
Rebased, added one test, made some logic changes to catch one other edge case and at the same time fixed the @hhovsepy issue.

**Comment by rhqci** [comment]
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#307](https://paste.fedoraproject.org/paste/BGYVm97-L97M0D7ocGwWHg)
  * :red_circle: Istio-and-Kiali-on-Openshift [#[1862]](https://paste.fedoraproject.org/paste/nEjtClp5ak1xUVuZn7SRlw)
  * :heavy_check_mark: Istio-Upstream [#[704]](https://paste.fedoraproject.org/paste/xIdh1msvxAgjOM3-GaDpsA)
  * :heavy_check_mark: Kiali-Istio-Checker [#[2264, 2266]](https://paste.fedoraproject.org/paste/oHUWfdwvuHAoHpOmTKGrGw)
  * :heavy_check_mark: install-bookinfo [#[1247]](https://paste.fedoraproject.org/paste/bWldzWW4jx3gMtfnnexdGQ)
  * :heavy_check_mark: Kiali-on-Openshift [#[1227]](https://paste.fedoraproject.org/paste/sMCdh2xiMObSGFt3tQnTWw)
  * :red_circle: run-kiali-e2e-tests [#[622]](https://paste.fedoraproject.org/paste/BqfYwGeLwKz7aTNI6Gg-0A)


KIALI_NAME=kialicorepr596-istio-system

**Comment by rhqci** [comment]
:heavy_check_mark: Jenkins CI: kiali-core-pr-e2e-test [#308](https://paste.fedoraproject.org/paste/tjyrmsGFrievwzy~bo01tQ)
  * :red_circle: Istio-and-Kiali-on-Openshift [#[1863]](https://paste.fedoraproject.org/paste/KuGydCAWZ-8C-bSofC77Vw)
  * :heavy_check_mark: Istio-Upstream [#[705]](https://paste.fedoraproject.org/paste/rsOyqbbrVwIPxtDS5QXuog)
  * :heavy_check_mark: Kiali-Istio-Checker [#[2267, 2268]](https://paste.fedoraproject.org/paste/CxaGb4IbUjzptGohK7dWuQ)
  * :heavy_check_mark: install-bookinfo [#[1248]](https://paste.fedoraproject.org/paste/CvX7bFKhkJDi9sNlRAZp-A)
  * :heavy_check_mark: Kiali-on-Openshift [#[1228]](https://paste.fedoraproject.org/paste/MeWu57wSMz~IkJmlGT4rMg)
  * :red_circle: run-kiali-e2e-tests [#[623]](https://paste.fedoraproject.org/paste/vLdODnXAqR2dTGVYk60fwA)


KIALI_NAME=kialicorepr596-istio-system

**Comment by hhovsepy** [comment]
Verified on
kiali-ui 0.10.0 (a4d7f884ddd333f29dd85ca210746a05f06cbe0f)
kiali v0.10.0-SNAPSHOT (ed514f2d985c8ad80b33ae01e98404a3d2b425f1)

**Comment by lucasponce** [comment]
sorry @burmanm, could you rebase when you have a moment ?
Thanks
